apollo-router/src/services/subgraph_service.rs

```diff
@@ -597,104 +593,205 @@ async fn call_http(
-                            status_code: Some(parts.status.as_u16()),
-                            reason: format!(
-                                "{}: {}",
-                                parts.status.as_str(),
-                                parts.status.canonical_reason().unwrap_or("Unknown")
-                            ),
-                        }))
-                    } else {
-                        Err(BoxError::from(FetchError::SubrequestHttpError {
-                            status_code: Some(parts.status.as_u16()),
-                            service: service_name.clone(),
-                            reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE}; found content-type: {content_type:?})", APPLICATION_JSON.essence_str()),
-                        }))
-                    };
-                }
-            }
-        }
+    // Perform the actual fetch. This
+    context.enter_active_request();
+    let (parts, body) = perform_fetch(client, service_name, request)
+        .instrument(subgraph_req_span)
+        .await
+        .map(|r| {
+            context.leave_active_request();
+            r
+        })?;
 
-        let body = match hyper::body::to_bytes(body)
-            .instrument(tracing::debug_span!("aggregate_response_data"))
-            .await {
-                Err(err) => {
-                    cloned_context.leave_active_request();
+    // Print out debug for the response
+    if display_headers {
+        tracing::info!(
+            http.response.headers = ?parts.headers, apollo.subgraph.name = %service_name, "Response headers from subgraph {service_name:?}"
+        );
+    }
 
-                    tracing::error!(fetch_error = format!("{err:?}").as_str());
+    if display_body {
+        if let Ok(body) = &body {
+            tracing::info!(
+                http.response.body = %String::from_utf8_lossy(body), apollo.subgraph.name = %service_name, "Raw response body from subgraph {service_name:?} received"
+            );
+        }
+    }
 
-                return Err(FetchError::SubrequestHttpError {
-                    status_code: None,
-                    service: service_name.clone(),
-                    reason: err.to_string(),
-                }.into())
+    // Parse the response.
+    // https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#processing-the-response
+    // If the response uses a non-200 status code and the media type of the response payload is application/json then the client MUST NOT rely on the body to be a well-formed GraphQL response since the source of the response may not be the server but instead some intermediary such as API gateways, proxies, firewalls, etc.
 
-                }, Ok(body) => body,
-            };
+    enum ContentType {
+        ApplicationJson,
+        ApplicationGraphqlJson,
+        Other,
+    }
+    let content_type = parts
```

Comment:
this was already extracted in `perform_fetch`, why not return it from that same function instead of looking at the header twice? That would avoid the `Other` element of the enum

Reply:
I've taken this into account in the rewrite.